core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/unmarshall/AwsXmlPredicatedResponseHandler.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.protocols.xml.internal.unmarshall;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.Response;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.SdkStandardLogger;
+import software.amazon.awssdk.core.exception.RetryableException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.core.http.HttpResponseHandler;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.protocols.query.unmarshall.XmlElement;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * Unmarshalls an HTTP response into either a successful response POJO, or into a (possibly modeled) exception based
+ * on a predicate that the unmarshalled response can be tested against. Returns a wrapper {@link Response} object which
+ * may contain either the unmarshalled success POJO, or the unmarshalled exception.
+ *
+ * @param <OutputT> Type of successful unmarshalled POJO.
+ */
+@SdkInternalApi
+public class AwsXmlPredicatedResponseHandler<OutputT> implements HttpResponseHandler<Response<OutputT>> {
+    private static final Logger log = LoggerFactory.getLogger(AwsXmlPredicatedResponseHandler.class);
+
+    private final Function<SdkHttpFullResponse, SdkPojo> pojoSupplier;
+    private final Function<AwsXmlUnmarshallingContext, OutputT> successResponseTransformer;
+    private final Function<AwsXmlUnmarshallingContext, ? extends SdkException> errorResponseTransformer;
+    private final Function<AwsXmlUnmarshallingContext, AwsXmlUnmarshallingContext> decorateContextWithError;
+    private final boolean needsConnectionLeftOpen;
```

Comment:
This seems out of place for this class...

Reply:
What seems out of place? The 'needsConnectionLeftOpen'? That is required because it's the responsibility of this class to close the connection.

Reply:
Why is that the responsibility of this class?

Reply:
Before I made this change it was already established as the responsibility of the 'success handler' to determine whether the underlying connection could/should be closed after handling the response, this is just a refactoring of that responsibility. In practice this flag should never be set to true because this class should not be used with streaming operations, but that's due to circumstance and I don't think that constraint should be codified.